### PR TITLE
Changes to support PP in megatron-rl

### PR DESCRIPTION
**Pull request by ArEsKay3**
# What does this PR do ?
Get pipeline parallel working in megatron rl loop.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by tdene** [comment]
@santhnm2 requesting your direct review

File: megatron/rl/rl_utils.py [hunk 1]
```diff
@@ -1616,33 +1616,64 @@ def prepare_data_for_update(
...
-                    b_attn_mask = None
 
-                logprobs = get_logprobs(
-                    model, b_trajs.cuda(), b_posids.cuda(), b_attn_mask, no_grad=True
-                )
-                old_logprobs.append(logprobs.detach().cpu())
+            if is_pipeline_last_stage():
+                old_logprobs = torch.concat(old_logprobs, dim=0)
+            else:
+                old_logprobs = torch.empty(len(compute_trajs), args.seq_length-1, dtype=torch.bfloat16, device=torch.cuda.current_device())
```
Comment (ArEsKay3) [inline megatron/rl/rl_utils.py]: This is a bug. I need to choose the dtype properly. Will dig to see how logprob dtype is determined. Or I could up/down cast...

File: megatron/rl/rl_utils.py [hunk 2]
```diff
@@ -1705,23 +1736,30 @@ def prepare_data_for_update(
...
+                if is_pipeline_last_stage():
+                    ref_logprobs.append(
+                        output_tensor[0].detach()
+                    )
 
-            ref_logprobs = torch.concat(ref_logprobs, dim=0)
+            if is_pipeline_last_stage():
+                ref_logprobs = torch.concat(ref_logprobs, dim=0)
+            else:
+                ref_logprobs = torch.empty(len(compute_trajs), args.seq_length-1, dtype=torch.bfloat16, device=torch.cuda.current_device())
```
Comment (ArEsKay3) [inline megatron/rl/rl_utils.py]: Same as above.

**Review by kvareddy — APPROVED** [review]

File: megatron/core/inference/contexts/dynamic_context.py [hunk 3]
```diff
@@ -385,7 +385,7 @@ def __init__(
 
         # Set max_total_requests, max_active_requests, max_tokens.
         self.max_total_requests = self.block_allocator.total_count - 1  # -1 for dummy block
-        self.max_active_requests = self.block_allocator.active_count
+        self.max_active_requests = math.floor(self.block_allocator.active_count / tp_size) * tp_size
```
Comment (yobibyte) [inline megatron/core/inference/contexts/dynamic_context.py]: nit: is this just integer division with //?

File: megatron/rl/rl_utils.py (hunk 1 above)
Comment (yobibyte) [reply megatron/rl/rl_utils.py]: Would args.bf16 or args.fp16 (otherwise fp32) work?

File: megatron/rl/inference/megatron.py [hunk 4]
```diff
@@ -233,7 +233,7 @@ async def launch(cls, model: GPTModel, **kwargs):
                            "wandb module is available. Inference logging will be disabled.")
 
         inference_engine: DynamicInferenceEngine = get_dynamic_inference_engine(args, model, inference_logging_step_interval, metrics_writer)
-        await inference_engine.start_listening_to_data_parallel_coordinator(inference_coordinator_port=41521, launch_inference_coordinator=True)
+        await inference_engine.start_listening_to_data_parallel_coordinator(inference_coordinator_port=41521, launch_inference_coordinator=True, verbose=False)
```
Comment (yobibyte) [inline megatron/rl/inference/megatron.py]: nit: Add an arg? Or todo to potentially move this to logging and later setup proper logging levels?

File: train_rl.py [hunk 5]
```diff
@@ -278,30 +282,36 @@ def forward_step(data_iterator, model: GPTModel, loss_only: bool = False):
 
     # Get current logprobs and calculate loss with straggler detection
     with stimer:
-        current_logprobs = get_logprobs(
+        logprobs_or_hidden_states = get_logprobs(
```
Comment (yobibyte) [inline train_rl.py:270]: Should we rename the get_logprobs function in this case?

File: megatron/rl/rl_utils.py [hunk 6]
```diff
@@ -1616,33 +1616,64 @@ def prepare_data_for_update(
...
                     )
-                    b_attn_mask = compute_attention_mask[start_idx:end_idx].cuda()
-                else:
-                    b_attn_mask = None
 
-                logprobs = get_logprobs(
-                    model, b_trajs.cuda(), b_posids.cuda(), b_attn_mask, no_grad=True
-                )
-                old_logprobs.append(logprobs.detach().cpu())
+            if is_pipeline_last_stage():
```
Comment (wdykas) [inline megatron/rl/rl_utils.py]: sorry I know this is kinda unrelated to this MR but can we try to start  using non-global parallel state for checks?  An example of the alternative for stage check is here https://github.com/NVIDIA/Megatron-LM/pull/2485/files#diff-7fc2a91343e49be3c1705b8237bbe437029132d8826e5c963e1b0b84d457bc38R914 and general use of the new process group is here https://github.com/NVIDIA/Megatron-LM/pull/2379/files#diff-47f76111bd3f07e1b77eedb8ae7b55fb00a8d8f089d1011b7633b257270f6d68R660. This  to will break things when we want run things with different parallelism

File: megatron/core/inference/sampling_params.py [hunk 7]
```diff
@@ -50,7 +50,7 @@ def _sync_prompt_logprobs_fields(self):
             assert (
                 self.skip_prompt_log_probs
             ), "return_prompt_top_n_logprobs requires skip_prompt_log_probs to be False"
-        if self.top_n_logprobs > 0:
+        if self.return_log_probs:
```
Comment (tdene) [inline megatron/core/inference/sampling_params.py:53]: @jalbericiola @ArEsKay3 please remove this particular 1-line diff.

**Comment by ArEsKay3** [comment]
Moved to #2632 Closing
